### ISSUE TITLE
Make Unifont and OpenMoji the default fonts

### ISSUE THREAD

**Issue by GrimPixel** [minetest/minetest#11455]
## Problem
People can't see some writing systems like Arabic in game, as well as emoji.

## Solutions
Make Unifont and OpenMoji the default fonts. Unifont supports as many writing systems as possible and is pixelated, suitable for the game's pixel art style; OpenMoji is an emoji font covering the latest Unicode version.

http://unifoundry.com/unifont/index.html
https://openmoji.org/

**Comment by hecktest** [comment]
It could be challenging to make a single pixel font work on every possible device and screen. It's also somewhat artistically imposing. An anti-aliased font like Arimo is as neutral as it gets, and at least guaranteed to render right if you throw enough DPI at it, without pixel perfect alignment.

The issues with rendering characters absent from the default font may be solved by using a fallback font to render those. For this it would be ideal to defer to what the operating system has to offer for a given language. That way you don't have to ship a font for every language in the world, including the latest corporate pictographic script. 🤮🤮🤮🤮🤮

**Comment by sfan5** [comment]
Font support isn't the primary issue, Minetest is missing support for anything right-to-left.
We might need https://github.com/fribidi/fribidi for the rendering at least.

**Comment by hlqkj** [comment]
I beg you, please don't let emojis invade minetest LOL

**Comment by LoneWolfHT** [comment]
> I beg you, please don't let emojis invade minetest LOL

If there was a setting like strip_color_codes that disabled emojis by default I'd be for it

**Comment by hlqkj** [comment]
Of course if that will ever be added a `strip_***` setting would be due. However, imho having such settings should not justify/allow excessive bloating of the engine with features like that: grant emojis, then why not stickers, sharing images, then idk what...

Minetest is a game engine and its chat is not supposed to be an instant messaging platform. Or is it?

**Comment by sfan5** [comment]
Minetest already automatically supports Emoji (in black-white) *if* the font contains the glyphs.

**Comment by hecktest** [comment]
But we don't need to go out of our way to provide the font. Let's not feed this beast.

**Comment by rubenwardy** [comment]
:'(

**Comment by SmallJoker** [comment]
ℑ 𝔡𝔬𝔫'𝔱 𝔰𝔢𝔢 𝔞𝔫𝔶 𝔟𝔢𝔫𝔢𝔣𝔦𝔱 𝔦𝔫 𝔈𝔪𝔬𝔧𝔦 𝔣𝔬𝔫𝔱𝔰. 𝔄𝔩𝔰𝔬 𝔄𝔯𝔞𝔟𝔦𝔠 𝔤𝔩𝔶𝔭𝔥𝔰 𝔞𝔫𝔡 𝔢𝔪𝔬𝔧𝔦𝔰 𝔞𝔯𝔢 𝔱𝔴𝔬 𝔢𝔫𝔱𝔦𝔯𝔢𝔩𝔶 𝔡𝔦𝔣𝔣𝔢𝔯𝔢𝔫𝔱 𝔯𝔢𝔮𝔲𝔢𝔰𝔱𝔰.
Arabic is also a blacklisted language, meaning that Minetest cannot render it properly. Duplicate of #4637

**Comment by Wuzzy2** [comment]
YES!! PLEASE do switch to a pixel font, it's long overdue. And GNU Unifont is quite nice, I am using it right now and I do not want to go back to the old font ever. The fact that the entire BMP is supported is also nice, this would instantly solve many problems with chat in other languages.

The font is something that has been woefully neglected in Minetest. Yet it is so important. It's a crucial part of a game's style. If text looks ugly, you don't want to read ...
Sans-serif doesn't seem to make sense most of the time. Minetest is based on the concept of voxels, which are just pixels in 3D. It is a no-brainer to me that the font is also pixel-style. Especially if the font is large, Unifont looks great. Minetest should look like a game and not like a Word document.

I don't get this argument of Minetest being supposed to look "neutral". I hate this argument, it is used again and again to shut down any artistic improvements which is one reason why Minetest's ugly style hasn't improved in years. There is no "neutral-looking" font, every font has to have some style, you can't get around that. So the best we can do is pick a font which best matches Minetest. And about "artistically imposing": So is the current default font. The default sans-serif font is artistically imposing on all games that want a pixel style. And considering how many games use 16×16 (i.e. pixelated) textures, I honestly do NOT understand the opposition to a pixel font.

Now I have to admit: Unifont also has a drawback: While it is a pixel font it is also a monospace font. **Please do not equate the two!** You can have a pixel font that isn't mono, you just need to have variable character length and it's no longer mono. Anyway, monospace might be annoying. A pixel font that isn't mono would also be nice. However, supporting all of the BMP is quite a killer argument IMO. Maybe Unifont could be the fallback font?

Maybe a long-term goal would be to provide a healthy set of fonts in Minetest by default and so games can pick and choose. The set of fonts should be based on what games need. Personally, I can tell youtaht MineClone 2's GUI looks horrible with the sans-serif font, and a pixel font is a MAJOR improvement. Once you activated the pixel font in MCL2, you never want to go back.
Oh, and this is also true for my other games.

Now, you might argue that it could be tricky to get it to work properly across devices and you might be right but that's not really a reason to reject it outright. I am pretty sure this is a solvable problem.

The OpenMoji part of the suggestion should be dropped tho. This is completely different from the question of using the default font. And IMO would be overkill. I don't like the OpenMoji suggestion, it is too heavy and not really needed IMO. Minetest is an engine for games, not an instant messenger. Besides, there are basic smiley faces in the BMP (as glyphs) which is good enough IMHO.

Summary: I fully support adding GNU Unifont to Minetest. And in the long run, Minetest could have a healthy selection of fonts made available for games.

**Comment by sfan5** [comment]
> The fact that the entire BMP is supported is also nice, this would instantly solve many problems with chat in other languages.

Which?
The mess we had where Chinese text would not show up if your language was not Chinese (plus the reverse in rarer cases) is no more. If you're thinking of Arabic, Hebrew, Hindi or Thai then AFAIK the problem is not font support but Minetest missing the required typesetting supporting since these writing systems are RTL and/or more complex.

**Comment by Wuzzy2** [comment]
Hmmm, I admit I don't know how good the coverage or the current default font is … But IIRC it is not the entire BMP, is it?

**Comment by sfan5** [comment]
No you're correct, our current fonts certainly don't cover the entire BMP. What I'm trying to say is that better font support doesn't buy us anything (at the moment) since the additional languages are not usable for other reasons.

**Comment by v-rob** [comment]
Here's what's wrong with our font support: We act as though Unicode is just UCS-2 or UCS-4, depending on platform.  This is __*not*__ Unicode.  Using UTF-32 is not Unicode either.  In specific, there's three things terribly wrong: first, on Windows, non-BMP text doesn't work right because we act as though a single 16-bit character is a whole character, which doesn't cover all of Unicode by far.  Second, we don't support combining characters (grapheme clusters), such as a combined `a` and accent.  Third, we don't support bidirectionality or complex layouts.

In other words, to fully support Unicode, we'll need to use a consistent encoding (no `std::wstring` at all, preferably UTF-8, since UTF-32 has no practical advantage) with a library that understands Unicode grapheme clusters (like [ICU](http://site.icu-project.org/)) and a library that can lay out complex languages properly (like [HarfBuzz](https://harfbuzz.github.io/)) and a library that can handle bidirectionality (like, again, ICU, or [GNU FriBiDi](https://github.com/fribidi/fribidi)).  In short, Unicode isn't easy, and there is no simple solution, least of all replacing the font.

---

More on-topic, I like the look of Unifont.  It's not a thick pixel font like Minecraft's, and it's not a bland sans-serif either.  I haven't actually tested it in game, but I'll bet it looks better.  On the other hand, I will not support an emoji font.

**Comment by GrimPixel** [comment]
Maybe find a method with FontForge to change monospace fonts to regular ones? FontForge supports Python scripts.

https://fontforge.org/
https://fontforge.org/docs/scripting/python.html

**Comment by hecktest** [comment]
Here's the problem though, we support various styles and sizes and ship with a variable width font by default.
![whatever](https://user-images.githubusercontent.com/42101236/126669136-4c80517f-be24-4e0a-a580-b2383f27209e.png)
![sizes](https://user-images.githubusercontent.com/42101236/126669142-54822855-bca7-4add-92a5-69cc2af54bcf.png)
I cannot imagine a pixel font in this situation looking like anything other than a horrible badly-scaled mess. I like the way my forms look right now, please don't pull this rug from under me too.

@Wuzzy2
That's a lot of words to say that _you_ like this font and you don't care how it would affect anyone else.
If you're so enthusiastic about this then _you_ should solve the problem of my forms looking like utter garbage after the hype fiends get this adopted, instead of just handwaving it as "solvable" (by whom I wonder).

Here, I'll even give you a demonstration:
![unifont_](https://user-images.githubusercontent.com/42101236/126672059-dbab871f-0a17-485d-b3f1-a3677a6addad.png)
![unifont](https://user-images.githubusercontent.com/42101236/126672069-c79e7d91-80a0-48fc-9df6-a214a5791e14.png)
![ugly](https://user-images.githubusercontent.com/42101236/126672075-a93d116a-feae-489a-9c96-92c694c2cdc9.png)
![awful](https://user-images.githubusercontent.com/42101236/126673976-5020c6ac-e610-4d58-bb79-2a996a077054.png)

I'll take a sans antialiased font over this blurry scaled pixel rubbish _any day_. This is a goddamn terminal font, you can put it in the console if you want. (since it's basically a worse Terminus)
It's actually really hard to read on top of a 3D background because of how thin the lines are:
![fastmode](https://user-images.githubusercontent.com/42101236/126672508-6c1414e2-071e-4dcb-ac53-6bf3482c95d2.png)
![garbage](https://user-images.githubusercontent.com/42101236/126672519-05849ba5-b54d-4138-8b25-f53d289ef7c6.png)
And look at how poorly it interacts with shadows.
![shadowgap](https://user-images.githubusercontent.com/42101236/126673359-62e03726-d51f-45af-bad2-efa92d25e2e6.png)

In what universe is an unscaleable, pixel thin font a good default?

**Comment by GrimPixel** [comment]
@hecktest then it's a problem of lacking customization of the font size, not the font. Even san-serif looks bad when stretched.

**Comment by rubenwardy** [comment]
Fonts are pretty important for overall game and feel. I'd like to see a fonts API, with a fallback to a font that supports most characters

**Comment by GrimPixel** [comment]
@Wuzzy2 No need to say so many “IMO”, I am already disheartened by the “pregnant man” emoji and don't want to talk about it anymore.

**Comment by v-rob** [comment]
hecks does seem to have somewhat of a point, Unifont doesn't seem to scale very well.  If we do have a font change, making sure it looks good in all situations is important (i.e. with a Minetest background, at different sizes, with bold and italic, etc).  However, I don't quite agree with hecks's "if we change the font it will break stuff, so don't change the font ever" (which at least seems to be what was coming across to me).

I do agree with rubenwardy that giving mods more power over fonts is a good idea.  It is, in fact, on my roadmap.  That is probably the best solution.

Of course, people seem to forget that pixel fonts have been around since forever.  It's called a font rasterizer without the ability to smooth fonts, like Windows pre-Vista:
![image](https://user-images.githubusercontent.com/31123645/126699532-deb47b4b-cb47-47ca-8753-70970c41d007.png)
It does scale properly, and works with anything, including TTF :)

**Comment by yw05** [comment]
I am not sure whether this is appropriate here, but as we are talking about fonts and character support:

Certain CJK characters that are written slightly differently depending on the country are unified into a single codepoint (even ArchWiki [has a section that addresses this issue](https://wiki.archlinux.org/title/Localization/Simplified_Chinese#Fixed_Simplified_Chinese_display_as_a_variant_(Japanese)_glyph), and I have experienced this myself with some signs IRL written in multiple languages). Maybe a "font selection formspec" that shows up when MT is run for the first time could help with that? (and it should hopefully settle the Unifont vs current font debate right now)

**Comment by GrimPixel** [comment]
I see people can scale the font based on the width of container with CSS easily. I suppose it is not hard to do here.

@yw05 Yes, some character variants share the same code in Unicode, not limited to CJK. Even Source Han Sans isn't perfect at distinguishing them. Users may be able to choose the font on their own computers, instead of having a “fonts” folder to be default. Some pixel font recommendations can be made.

**Comment by Wuzzy2** [comment]
Strange, I didn't have these broken characters for any font size that I tried. Is this a bug that only occurs for some?

Also, I'm not saying Unifont should beome the ONLY font remaining. It would be best if games had a choice between several high-quality fonts.

**Comment by hecktest** [comment]
@GrimPixel 
> let's see if your troubles with a pixel font really exist

What are you talking about dude, I posted actual screenshots there. It's not a contrived example, I literally downloaded Unifont TTF, set it in the config and ran my game. We have styling in formspecs now and Unifont just straight up doesn't scale. I used `<style size=12>` and `<big>` respectively in the two examples that scale poorly.

Actually, you seem to have a little problem here...
![HqHtkZFD](https://user-images.githubusercontent.com/42101236/126721100-cbfdaebd-17a1-4987-a3de-1b9cb3c84d45.png)
![wGXfSXjM](https://user-images.githubusercontent.com/42101236/126721145-e8b27c1d-8018-45bb-9275-959e828d0946.png)

Also, the font is 12MB, that's just decadent.

@v-rob 
I still wouldn't take pixel Tahoma over an anti-aliased Arial clone, even if it magically provided all the sizes. Don't forget that the display tech itself has changed since, a pixel font on a CRT looks different from one on an LCD (which has orthogonal gaps between the dots that affect how lines are perceived; pixels feel a lot sharper and non-ortho lines are harder on the eyes, also color visibly affects horizontal alignment where on a CRT it would be somewhat randomly distributed)

**Comment by v-rob** [comment]
Non-smoothed fonts was mostly a joke.

However, I don't agree that changing the font should be a breaking change.  If changing the font will break your formspecs, I'm willing to bet that resizing the window will break your formspecs.  Fonts don't resize with the window (they're based on pixels), but formspec coordinates do (they are not always 60 pixels, but a calculation based on the window dimensions).

Secondly, even if your formspecs do break only if the font is changed, the user can still change the font in the settings.  In fact, making your fonts absolute sized is bad for accessibility (there's a reason CSS `px` is highly discouraged for fonts, and `em` highly encouraged).  All _reasonable_ user preferences should be accommodated for, so if a player needs a slightly larger font to see better, they shouldn't be overridden.  I never should have implemented absolutely sized fonts if they aren't used responsibly.

**Comment by hecktest** [comment]
Resizing the window doesn't break my forms, they do not scale. Either I'm flipping some flag on those to make that happen or they just behave that way.

I assume a minimum resolution of 1280x720/96 (if that wasn't apparent from the template) and my forms work on that resolution and above.

I also assume that DPI changes translate to a uniform scale of all GUI elements. Gonna test that in a moment, but I'll have to relog since MT seems to just obey the system DPI.

**Comment by hecktest** [comment]
Rendering at 150% DPI confirms my hunch. It's all scaled uniformly and doesn't break the layout. I would recommend just using this for accessibility rather than blowing up the font without the room to render it. This is also crucially a feature that would be completely borked on a pixel font, so that's one more point for keeping a vector sans font as the default.

Shame that it breaks my pixel perfect GUI sprites but you can't really do anything about that, it's fair enough for an accessibility option. A 1440p screen could render it at 200% DPI which would look right once again.

![creator](https://user-images.githubusercontent.com/42101236/126729643-8362bdc2-20ba-481d-8ba5-75a1e5fedcfb.png)

**Comment by GrimPixel** [comment]
@hecktest Oh, you forgot to make it a comparison, let me help you:
![MT default font enlarged 1](https://user-images.githubusercontent.com/39795763/126751608-10455cc6-6fce-4200-8872-211a0f654596.png)
https://user-images.githubusercontent.com/42101236/126721960-48893735-526a-48d0-9c05-f10768aa4768.png

![MT default font enlarged 2](https://user-images.githubusercontent.com/39795763/126751612-0db192bd-e1f0-472e-a7d6-260c78595885.png)
https://user-images.githubusercontent.com/42101236/126721950-c8e711f1-6fa3-4c05-9285-a84a79c0a3d4.png

The default fonts folder is 13.8 MB, has bold and italic, which is good and can be used in cases I've not experienced yet, but the supported number of writing systems is still a problem.

As there is no perfect font catering for everybody, why not simply let the user choose from their installed fonts in settings?

**Comment by Wuzzy2** [comment]
WTF is going on with the 0 in the Unifont? This must be a bug, right?

But yeah, I agree the default font for GAMES should better not be touched. Instead, just add more font options for games to choose from but they only activate when explicitly requested. Hecktest is 100% correct that "pulling the rug under you" is naughty and shouldn't be done unless there's a very good reason. Unifont is not a very good reason.
It seems that Minetest devs have been too bold in changing behavior in a non-backwards-compatible way in the recent past. The respect for backwards-compability needs to be reinstated.

I still think it's worthwhile to have SOME pixel font included. It clearly would benefit games like MCL2 a lot. I still was unable to reproduce those silly artifacts tho. o_O

The file size of the font is of no concern to me. Music and sounds are what really increase game size (and hi-res textures).

**Comment by sfan5** [comment]
> The default fonts folder is 13.8 MB, [...]

It's actually only 5.5 MB, the `xml` and `png` files do not get installed.

> why not simply let the user choose from their installed fonts in settings?

You can already do this, Minetest just has no knowledge of where it'd find "installed" fonts so you have to specify a path.

**Comment by hecktest** [comment]
@GrimPixel 
I don't know what you intended to show us by this comparison, because it shows Arimo rendering as intended and Unifont being utterly broken.

Look at the halos around the Unifont glyphs, this is a scaling error.
Using non-integer scales defeats the purpose of using a pixel font.

Curiously, I do not get these halos:
![nohalos](https://user-images.githubusercontent.com/42101236/126776423-6ffa35f8-0538-46b9-939d-c0439a372a62.png)
This is probably because we're running different DPI settings. MT defaults to 72 but on Windows it uses the system setting where the minimum is 96, and it seems 96 is the spot where font pixel sizes actually correspond to what's being displayed on the screen.

Here are the more obviously broken glyphs from your screenshots:
![broken](https://user-images.githubusercontent.com/42101236/126777505-84bff2ec-8b29-4964-bcea-870999115cf6.png)

Same thing happens in my example whenever I deviate from a 16px font size. Notice how only one of these breaks.
![compare](https://user-images.githubusercontent.com/42101236/126777271-45e362aa-de11-45ca-9fb0-688448a8d892.png)

Bottom line is, we support font styling and DPI scaling and it's difficult to make a pixel font work in these conditions; the font will break more often than it will work correctly. Changing the default from a vector sans font at this point would be regressive.

By the way, as much as @v-rob calls absolute sizing a mistake, **it's actually necessary for pixel fonts to work at all**. If you want a pixel font for Mineclone or another game, you _need_ absolute font sizes and server-sent fonts so that you actually have some control over how things are rendered!

**Comment by v-rob** [comment]
Yes, absolute font sizes are necessary for pixel fonts.  With the evidence of pixel fonts breaking at different sizes, I certainly do not support them.  Unless there is a proper way to address these issues, I wouldn't support adding a pixel font, much less as the default font.

@hecktest: I believe I might know why your forms do not scale.  In your forms, do you have use `size[#,#,true]`, as opposed to `size[#,#]` or `size[#,#,false]`?  That's the only thing I can think of that would cause formspec coordinates to be a fixed size.  However, on 96 DPI, it should result in 53 pixel coordinates, not 60.  If your coordinates are fixed without using `size[#,#,true]`, that's a bug.

Note that I don't think that fixing coordinate size is particularly good practice, but I suppose that if it works and people use it, we have to support it.

**Comment by hecktest** [comment]
Yup, it appears that I made all my forms with fixed=true. This must have been an intentional decision, I just don't remember it. I think this is the only sane formspec workflow if you want to do anything nontrivial.

You seem to consistently hate the idea of using predictable, relatively fixed sizes for things, but likewise I don't understand this ideal of flexing and stretching and squashing every single UI element at will; that kind of system sounds like a complete nightmare to design anything for.
Between arbitrarily resizing random elements and just DPI scaling the whole thing, I think the latter is more practical.

**Comment by v-rob** [comment]
Have you developed anything in any GUI toolkit under the sun?  They have things known as sizers, grids, flex, relative dimensioning and similar because different people have differently sized screens.  What fixed sizes work on Windows will not work under Android, and what works on a normal laptop screen will not work on an ultra-large screen.  Having elements resize on different screen sizes/DPI is not "not practical", but the standard thing to do absolutely everywhere.  I don't claim that formspecs do it in a good way (they don't, and I can understand why fixed sizes are easier), but a fixed size is worse, and your formspecs will only work well on the screen sizes you've developed them on.

But whatever.  You can develop your programs in a brittle way that can easily break, and I can't force you to do otherwise.  However, bear in mind that the formspec replacement will not have any fixed sizes at all, because I'm trying to make something that actually works well everywhere.

**Comment by hecktest** [comment]
@v-rob 
Have you used any game engine under the sun? They usually solve these things in terms of anchors and uniform scaling, while keeping element sizes and positions fixed relative to one another where needed - because video games are not HTML pages or word processors. I don't think I've seen a single game UI out there doing this sort of thing, and not a single game UI developer ever in the universe has had to deal with something as ridiculous as _not knowing which font and size they're designing for_!

Try actually making something compact, pleasing and nontrivial in the forms system, using the new styling features, _without_ leaving tons of padding around every element to accomodate for stretching, and without ugly scrollbars everywhere.

I try to do my best with the tools I have, being practical about it, and you're just being condescending for no reason. If your new system won't allow a consistent workflow then it will literally be inferior to formspecs.

**Comment by erlehmann** [comment]
As @hecktest showed in this thread, pixel fonts are simply not viable as TTFs.

I want to repeat here that I want to help integrate GNU Unifont. But for that to not look ugly non-TTF support needs to come back:
https://github.com/minetest/minetest/issues/11611#issuecomment-921428980

**Comment by erlehmann** [comment]
@ExeVirus by “exposing DPI”, do you mean screen DPI or font DPI?

In case you mean font DPI: AFAIK Minetest does not actually use that for font rendering.

In case you mean screen DPI: Do you mean real physical screen DPI or the fake 96×96 X11 thing?

**Comment by sfan5** [comment]
By "exposing DPI" people mean exposing the value Minetest uses to scale all gui elements and font sizes, regardless of how that values came to be.
